# Домашнє завдання до теми "Розширені методи обробки даних у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('data/supermarket_sales.csv')

1. Відсортуйте дані за спаданням ціни товара ('Unit price') і виведіть 5 перших рядочків на екран.

In [3]:
df.sort_values(by='Unit price', ascending=False).head()

Invoice ID Branch       City Customer type Gender         Product line  \
983  148-41-7930      C  Naypyitaw        Normal   Male    Health and beauty   
122  219-22-9386      B   Mandalay        Member   Male    Sports and travel   
930  641-62-7288      B   Mandalay        Normal   Male   Home and lifestyle   
494  437-53-3084      B   Mandalay        Normal   Male  Fashion accessories   
283  667-92-0055      A     Yangon        Member   Male    Health and beauty   

     Unit price  Quantity  Tax 5%    Total       Date   Time      Payment  \
983       99.96         7  34.986  734.706  1/23/2019  10:33         Cash   
122       99.96         9  44.982  944.622   3/9/2019  17:26  Credit card   
930       99.92         6  29.976  629.496  3/24/2019  13:33      Ewallet   
494       99.89         2   9.989  209.769  2/26/2019  11:48      Ewallet   
283       99.83         6  29.949  628.929   3/4/2019  15:02      Ewallet   

       cogs  gross margin percentage  gross income  Rating  
983  699.72                 4.761905        34.986     6.1  
122  899.64                 4.761905        44.982     4.2  
930  599.52                 4.761905        29.976     7.1  
494  199.78                 4.761905         9.989     7.1  
283  598.98                 4.761905        29.949     8.5

2. Знайдіть інвойс (рядок в даних) в категорії товарів `Electronic accessories`, в якому був куплений товар з найбільшою кількістю і найбільшою ціною. Виведіть цей рядок даних на екран.

In [8]:
filter_by_product_line = df['Product line'] == 'Electronic accessories'
df[filter_by_product_line].sort_values(by=['Quantity', 'Unit price'], ascending=[False, False]).head(1)

Invoice ID Branch    City Customer type Gender            Product line  \
105  704-48-3927      A  Yangon        Member   Male  Electronic accessories   

     Unit price  Quantity  Tax 5%    Total       Date   Time  Payment   cogs  \
105       88.67        10  44.335  931.035  1/12/2019  14:50  Ewallet  886.7   

     gross margin percentage  gross income  Rating  
105                 4.761905        44.335     7.3

3.1. Створіть нову колонку в даних `Rating Category`, яка визначає категорію рейтингу (`Rating`): високий, середній чи низький за наступним правилом:
- якщо рейтинг вищий або рівний 8 - то він високий (`High`)
- якщо рейтинг від 6 до 8 не включаючи - то він середній (`Medium`)
- інакше - низький (`Low`).

In [9]:
def rating_category(value):
    if value >= 8:
        return 'High'
    elif value > 6 and value < 8:
        return 'Medium'
    else:
        return 'Low'

In [10]:
df['Rating Category'] = df['Rating'].apply(rating_category)

In [11]:
df[['Rating Category', 'Rating']].head()

Rating Category  Rating
0            High     9.1
1            High     9.6
2          Medium     7.4
3            High     8.4
4             Low     5.3

3.2. Виведіть, який відсоток даних становить кожне зі значень категорій рейтингу.

In [12]:
df['Rating Category'].value_counts(normalize=True)

Rating Category
Low       0.341
Medium    0.330
High      0.329
Name: proportion, dtype: float64

4.1. Запустіть код нижче аби створити змінну `time`. Ця змінна містить час в форматі "година:хвилина". Напишіть функцію `get_hour`, яка буде зі змінної формату як `time` дістати лише години і конвертувати їх у тип `int`.

Запустіть написану функцію на змінній `time` та перевірте правильність роботи функції.

In [13]:
time = '11:28'

In [20]:
def get_hour(time_str):
    hour_str = time_str.split(":")[0]
    return int(hour_str)

In [21]:
get_hour(time)

11

4.2. З допомогою функції `get_hour` і метода `DataFrame.apply()` створіть нову колонку `Hour` в нашому наборі даних, яка буде містити години здійснення покупки, обчислені на основі часу здійснення покупки `Time`.

In [22]:
df['Hour'] = df['Time'].apply(get_hour)

In [28]:
df[['Hour','Time']].head()

Hour   Time
0    13  13:08
1    10  10:29
2    13  13:23
3    20  20:33
4    10  10:37

4.3. Виведіть кількість значень за кожною годиною покупки і дайте відповідь "О яких годинах найбільше здійснюють покупок покупці і чому на Ваш погляд саме о цих годинах? Чи це слогічно, що найбільше покупок саме в ці години?".

In [29]:
df['Hour'].value_counts()

Hour
19    113
13    103
15    102
10    101
18     93
11     90
12     89
14     83
16     77
20     75
17     74
Name: count, dtype: int64

Найбільша кількість покупок робиться о 10, 13, 15 і 19 годині. Все логічно.   
О 10 годині - ранкові покупки, магазини тільки но відкрились, немає натовпів та черг.  
О 13 годині - обідня перерва.  
О 15 годині - батьки роблять покупки, забираючи дітей зі школи. Люди, що працються в декілька змін звільняються з роботи. Також скупляються на вечерю.  
О 17 годині - вечірні покупки, часто одразу після роботи.

In [32]:
df['Product line'].unique()

array(['Health and beauty', 'Electronic accessories',
       'Home and lifestyle', 'Sports and travel', 'Food and beverages',
       'Fashion accessories'], dtype=object)

5. Виведіть середнє значення чеку (`Total`) для покупок в містах, наявних в даних, за допомогою функції `groupby`.

In [34]:
df.groupby('City')['Total'].mean()

City
Mandalay     319.872506
Naypyitaw    337.099715
Yangon       312.354031
Name: Total, dtype: float64

6. Виведіть кількість покупок, середнє, мінімальне та максимальне значення для загальної суми чеку (Total) в містах, наявних в даних, за допомогою функції groupby.
Зробіть висновки з цих показників та дайте рекомендації для мережі супермаркетів.

In [38]:
df.groupby('City')['Total'].agg(['size', 'mean', 'min', 'max'])

size        mean      min      max
City                                         
Mandalay    332  319.872506  18.6375  1022.49
Naypyitaw   328  337.099715  10.6785  1042.65
Yangon      340  312.354031  12.6945  1039.29

- Висновки:  
    - В місті Yangon найменший середній чек 312.35 та найбільша кількість транзакцій 340, що свідчить про частіші покупки, але менш дорогі.
    - Для міста Mandalay середній чек 319.87, що більше, ніж у Yangon. Кількість покупок середня серед усіх міст, а найдешевша покупка в цьому місті 18.64 - це найбільше значення, серед трьох міст. Це вказує на потенціал для зростання прибутку.
    - В місті Naypyitaw найбільший середній чек 337.1 з найменшою кількістю покупок 328. Маємо меншу кількість клієнтів, що втрачають більше.
- Рекомендації для підвищення прибутку:
    - В Yangon стимулювати більші чеки, для цього запропонувати бонуси або знижки для більшої кількості товарів у чеку. Запустити рекламну кампанію. Покращити видимість товарів в магазині з вищою маржинальністю.
    - В місті Mandalay збільшити кількість покупок. Для цього активізувати програму лояльності, щоб покупки робили частіше.
    - В місті Naypyitaw дослідити поведінку покупців: які категорії товарів більше купують, яким способам розрахунку віддають перевагу, чи є переважна більшість покупців членами клубу. І використати ці інсайти в інших магазинах.

7. Ми внесли деякі зміни в наш набір даних, тож, давайте запишемо результат у файл. Запишіть оновлену версію даних у файл `supermarket_sales_enriched.csv`. Зверніть увагу, аби не записати нічого зайвого, наприклад, індекси рядків нам в файлі не потрібні, оскільки вони стандартні.

In [40]:
df.to_csv('supermarket_sales_enriched.csv', index=False)